docs/getting_started_jax.ipynb

```diff
@@ -0,0 +1,669 @@
+        "outputId": "4cc86102-6f47-44c0-ebac-171f55c1042e"
+      },
+      "outputs": [
+        {
+          "name": "stdout",
+          "output_type": "stream",
+          "text": [
+            "module @jit_body attributes {mhlo.num_partitions = 8 : i32, mhlo.num_replicas = 1 : i32} {\n",
+            "  func.func public @main(%arg0: tensor<32x64xf32>) -> (tensor<32x64xf32> {jax.result_info = \"\"}) {\n",
+            "    %0 = stablehlo.custom_call @Sharding(%arg0) {backend_config = \"unspecified_dims=[0,1]\", mhlo.sharding = \"{devices=[4,1,2]<=[8] last_tile_dim_replicate}\"} : (tensor<32x64xf32>) -> tensor<32x64xf32>\n",
+            "    %1 = stablehlo.custom_call @SPMDFullToShardShape(%0) {backend_config = \"unspecified_dims=[0,1]\", mhlo.sharding = \"{devices=[1,1,4,2]<=[8] last_tile_dims={manual, replicated}}\"} : (tensor<32x64xf32>) -> tensor<8x64xf32>\n",
+            "    %2 = call @shmap_body(%1) : (tensor<8x64xf32>) -> tensor<32x64xf32>\n",
+            "    %3 = stablehlo.custom_call @Sharding(%2) {backend_config = \"unspecified_dims=[0,1]\", mhlo.sharding = \"{devices=[1,1,4,2]<=[8] last_tile_dims={manual, replicated}}\"} : (tensor<32x64xf32>) -> tensor<32x64xf32>\n",
+            "    %4 = stablehlo.custom_call @SPMDShardToFullShape(%3) {backend_config = \"unspecified_dims=[0,1]\", mhlo.sharding = \"{replicated}\"} : (tensor<32x64xf32>) -> tensor<32x64xf32>\n",
+            "    return %4 : tensor<32x64xf32>\n",
+            "  }\n",
+            "  func.func private @shmap_body(%arg0: tensor<8x64xf32>) -> (tensor<32x64xf32> {jax.result_info = \"[None, None]\"}) {\n",
+            "    %0 = stablehlo.custom_call @Sharding(%arg0) {backend_config = \"unspecified_dims=[1]\", mhlo.sharding = \"{devices=[2,1,4]<=[4,2]T(1,0) last_tile_dims={manual}}\"} : (tensor<8x64xf32>) -> tensor<8x64xf32>\n",
+            "    %1 = \"stablehlo.all_gather\"(%0) <{all_gather_dim = 0 : i64, channel_handle = #stablehlo.channel_handle<handle = 1, type = 1>, replica_groups = dense<[[0, 2, 4, 6], [1, 3, 5, 7]]> : tensor<2x4xi64>, use_global_device_ids}> : (tensor<8x64xf32>) -> tensor<32x64xf32>\n",
+            "    return %1 : tensor<32x64xf32>\n",
+            "  }\n",
+            "}\n",
+            "\n",
+            "None\n"
+          ]
+        }
+      ],
+      "source": [
+        "def run_shard_map():\n",
+        "  x = jax.ShapeDtypeStruct((32, 64), jnp.float32)\n",
+        "\n",
+        "  def body(x):\n",
+        "    x = jax.lax.with_sharding_constraint(x, NamedSharding(mesh, PartitionSpec('model', PartitionSpec.UNCONSTRAINED)))\n",
+        "    return jax.lax.all_gather(x, 'data', tiled=True)\n",
+        "\n",
+        "  shmaped_f = shard_map(\n",
+        "        body,\n",
+        "        mesh=mesh,\n",
+        "        in_specs=(jax.sharding.PartitionSpec('data',),),\n",
+        "        out_specs=jax.sharding.PartitionSpec(),\n",
+        "        check_rep=False,\n",
+        "        auto = frozenset(['model']))\n",
```

Comment:
After checking with JAX folks, so JAX doesn't have any public documentation on `auto = frozenset` yet, so we should remove this argument for now. And then also could you re-run this cell to make sure the output is the latest? And re-run the cell below with this in it as well for the same reason?
```
        "jax.config.update(\"jax_use_shardy_partitioner\", True)\n",
        "run_shard_map()\n",
        "jax.config.update(\"jax_use_shardy_partitioner\", False)"
```

Reply:
I've hit a problem here, since removing the `auto=frozenset(['model'])` argument causes an error:

<details>

```
---------------------------------------------------------------------------

JaxStackTraceBeforeTransformation         Traceback (most recent call last)

[/usr/lib/python3.10/runpy.py](https://localhost:8080/#) in _run_module_as_main()
    195         sys.argv[0] = mod_spec.origin
--> 196     return _run_code(code, main_globals, None,
    197                      "__main__", mod_spec)

33 frames

JaxStackTraceBeforeTransformation: ValueError: Axis: model of PartitionSpec('model', UNCONSTRAINED) is also found in manual_axes: frozenset({'data', 'model'}).

The preceding stack trace is the source of the JAX operation that, once transformed by JAX, triggered the following exception.

--------------------


The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)

    [... skipping hidden 21 frame]

    [... skipping hidden 2 frame]

[/usr/local/lib/python3.10/dist-packages/jax/_src/sharding_impls.py](https://localhost:8080/#) in _check_mesh_resource_axis(mesh, parsed_pspec, _manual_axes)
     64               f"is not found in mesh: {tuple(mesh.shape.keys())}.")
     65         if r in _manual_axes:
---> 66           raise ValueError(
     67               f"Axis: {r} of {parsed_pspec.get_partition_spec()} "
     68               f"is also found in manual_axes: {_manual_axes}.") from None

ValueError: Axis: model of PartitionSpec('model', UNCONSTRAINED) is also found in manual_axes: frozenset({'data', 'model'}).
```

</details>

Do you have suggestions on how to work around this error?

I have also found this: https://jax.readthedocs.io/en/latest/notebooks/shard_map.html#api-specification

I know it probably doesn't count as acual docs, but it does mention the auto option (it is also described correctly in the API reference for `shard_map` https://jax.readthedocs.io/en/latest/_autosummary/jax.experimental.shard_map.shard_map.html

Reply:
Oh sorry missed this! Ah update the example to 
```py
def body(x):
    return jax.lax.all_gather(x, 'data', tiled=True)

  shmaped_f = shard_map(
        body,
        mesh=mesh,
        in_specs=(jax.sharding.PartitionSpec('data',),),
        out_specs=jax.sharding.PartitionSpec(),
        check_rep=False)
```
aka removing the `jax.lax.with_sharding_constraint`

Reply:
Done - thanks!